Change column type

pavish opened the pull request:
Fixes #498 

This PR builds on top of #627, containing commits from @asharonbaltazar and @dmos62.
This PR additionally handles:
* Displaying and allow choosing db types. (Previous PR only handled mathesar types).
* Refactoring related to the recent store structure.
* Reorganization of table header related files.
* Types store rewrite to preserve data flow
* Preload types on initial load

Demo video:
<a href="https://www.loom.com/share/fca85fdb7579429c92d820f7ea813e9f">
    <p>Column type change demo - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/fca85fdb7579429c92d820f7ea813e9f-with-play.gif">
  </a>

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

kgodey commented:
@pavish I watched the video. I think it looks good overall, but I have some concerns about the UX, it's diverging significantly from what we have in the data types design. In the design, we don't support changing the database type, we guide the user through it by asking them questions. We do show the database type that we've calculated based on their answers at the bottom, but it's not able to be selected or changed.

As for the questions specific to each data type, we need to resolve #661 before we can store all that information on the backend. I'm not sure that we'll be able to get to that any time soon since we're behind on data types implementation in the backend and I think we should prioritize getting the basic implementation done. I think it makes more sense to hardcode in the frontend for now and refactor later. However, that doesn't need to be handled in this PR, we can implement each one in the data types milestone.

seancolsen commented on mathesar_ui/src/sections/table-view/header/header-cell/type-options/TypeOptions.svelte:
```diff
@@ -0,0 +1,151 @@
+    <!-- TODO: Make tab container more generic to be used here -->
+    <ul class="type-option-tabs">
+      <li>
+        <Button appearance="ghost" class="padding-zero type-option-tab">
+          <Icon size="0.75em" data={faDatabase}/>
+          <span>Database</span>
+        </Button>
```
Using a `Button` seems strange here because it gives a `pointer` cursor but has no function when clicked.

pavish replied:
There is a comment above that block which says: `TODO: Make tab container more generic to be used here 
`

As per [our UX](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?node-id=4270%3A39549&starting-point-node-id=4270%3A39549), this should be a tab with a couple options. However, our current Tab component is not generic enough to handle this. I've used a button for the time being.

As we discussed in one of our previous conversations, we will have to create a generic Tab component in the library and a mathesar specific one which builds over it. We should be able to use the generic Tab component here.

seancolsen reviewed (changes requested):
## General notes

- Since this PR is big and old and contains work from multiple people, I would err on the side of getting it merged soon and then polishing it as-needed in subsequent PRs.
- For the above reason, I've not read the code as closely as I might otherwise. I did interact with the feature though.

## I like

- Changes to the column header menu, like triggering on the entire header, and adding grouping.

## I suggest

- Fix bugs:

    - Opening a second tab wipes out all the column type info. I can provide more specific replication steps if needed.
    - Uncaught error when trying to change column type of PK to text

- To somewhat piggyback on Kriti's comment, I'm concerned that presenting "Column Type" _and_ "Type in db" choices will be confusing to most users. I think we should present one flat list of types for the user to choose from. So if we want to allow the user to choose between db types `TEXT` and `VARCHAR`, then we should come up with two distinct "Column Type" options for those db types. NocoDB appears to do this by having types called "SingleLineText" and "LongText" which seem to be decent names.

kgodey commented:
From @seancolsen's comment:

>  So if we want to allow the user to choose between db types TEXT and VARCHAR, then we should come up with two distinct "Column Type" options for those db types. 

We do not want this. We have the UX/design for DB types completed – I'm not sure why it was implemented differently in this PR. @pavish, I assume you ran into some sort of issue with the design as specced. It would be useful to discuss design problems that you run into while implementing before you make changes so that we can be on the same page about the solution (and you won't have to undo work).

pavish commented:
@kgodey @seancolsen 

>  I'm not sure why it was implemented differently in this PR.

In our [UX/design for DB types](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?node-id=4270%3A39549&starting-point-node-id=4270%3A39549), we have an additional select menu to choose subtype for some abstract types. For eg., Number has "Integer, Decimal, Float". Based on the selected subtype and some additional configurations, we determine the db type.

We are essentially placing the frontend in control of identifying the db type since all the requests from the frontend use the db type and not the abstract types. The specifications on how to determine the db type from the abstract type + configurations are yet to be detailed for all of these abstract types. This determination step is unique to each abstract type, and needs to be done on the Data types milestone.

In this PR, I needed to allow the user to change the column type but since we do not know the specifications on how to determine the db type, I've allowed directly choosing it. The structure isn't very different from our UX and once we know the subtypes for each abstract type, we can just rename the text and use the same Select menu for the sub types. In cases where there are no subtypes, we'll just hide the Select menu.

> It would be useful to discuss design problems that you run into while implementing before you make changes so that we can be on the same page about the solution

This isn't a design issue but rather a scoping issue of this ticket. We cannot send a request to the server without knowing the db type, and we cannot know the db type without specifying how to do identify it, which is covered individually in the data types milestone. If I had not implemented choosing the db type, this whole feature would have been non-functional until the data types milestone is completed.

pavish commented:
@seancolsen 
> Opening a second tab wipes out all the column type info. I can provide more specific replication steps if needed

Great catch! I added a commit to fix that: https://github.com/centerofci/mathesar/pull/786/commits/6afcece04fad3f139d0872aef6ccaf1138bb956e

> Uncaught error when trying to change column type of PK to text

This was intentional since we hadn't decided on where to place the error boundaries for the toast system. For now, I've added a commit https://github.com/centerofci/mathesar/pull/786/commits/57ce2e75664a08d2736721c95af1ca84ab54af82 which catches it on the component. We could do it on the store, or even within the api layer of the store based on the pattern we discussed. This needs to be decided when we take up the toast system.

pavish commented on mathesar_ui/src/stores/abstractTypes.ts:
```diff
@@ -0,0 +1,177 @@
+  }
+}
+
+function processTypeResponse(abstractTypeReponse: AbstractTypeResponse[]): AbstractTypeStoreData {
```
Thanks. I've updated this in https://github.com/centerofci/mathesar/pull/786/commits/f1dfed429c457130af0a48d64eeeca529cee5959

pavish commented on mathesar_ui/src/stores/abstractTypes.ts:
```diff
@@ -0,0 +1,177 @@
+  state: States,
+  data: AbstractTypeStoreData,
+  error?: string
+}
```
Makes sense.. I have made some changes in https://github.com/centerofci/mathesar/pull/786/commits/f1dfed429c457130af0a48d64eeeca529cee5959.

I've introduced a pattern which calls the content of a store as Substance. Based on this:
* The type that implements `Writable` or `Readable` would be suffixed by `Store`.
* The value that is present within a store would be suffixed by `Substance`.

In this case, I've updated it to:
```typescript
export type AbstractTypesMap = Map<AbstractType['identifier'], AbstractType>;

interface AbstractTypesSubstance {
  state: States,
  data: AbstractTypesMap,
  error?: string
}

// Not in code, just for representational purpose
type AbstractTypesStore = Writable<AbstractTypesSubstance>;
```

Let me know if this looks good, or if we can come up with a better term instead of `Substance`.

seancolsen replied:
> calls the content of a store as Substance

I like it!

pavish commented on mathesar_ui/src/stores/abstractTypes.ts:
```diff
@@ -0,0 +1,177 @@
+}
+
+const databasesToAbstractTypesStoreMap: Map<Database['id'], Writable<AbstractTypeStore>> = new Map();
+const dbMTTypesRequestMap: Map<Database['id'], CancellablePromise<AbstractTypeResponse[]>> = new Map();
```
Previously `AbstractTypes` were called `MathesarTypes`. This lead to a lot of readability issues on the frontend and I renamed it to `AbstractTypes` which made more sense.

`MT` here referred to `MathesarType`. I missed this line when I updated the term. I've updated it in https://github.com/centerofci/mathesar/pull/786/commits/f1dfed429c457130af0a48d64eeeca529cee5959.

seancolsen reviewed (approved):
Thanks for addressing my feedback @pavish 

While there's still some work to do here, I'd like to get this merged so that we can begin to base other PRs off this one.

@kgodey do you have concerns that still need to be addressed before we merge?
